src/Compilers/Core/Portable/Collections/Rope.cs

```diff
@@ -0,0 +1,151 @@
+                r2.Length == 0 ? r1 :
+                checked(r1.Length + r2.Length < 32) ? ForString(r1.ToString() + r2.ToString()) :
+                new ConcatRope(r1, r2);
+        }
+
+        /// <summary>
+        /// Two ropes are "the same" if they represent the same sequence of characters.
+        /// </summary>
+        public override bool Equals(object obj)
+        {
+            if (!(obj is Rope other) || Length != other.Length)
+                return false;
+            if (Length == 0)
+                return true;
+            var chars0 = GetChars().GetEnumerator();
+            var chars1 = other.GetChars().GetEnumerator();
+            while (chars0.MoveNext() && chars1.MoveNext())
+            {
+                if (chars0.Current != chars1.Current)
+                    return false;
+            }
+
+            return true;
+        }
+        public override int GetHashCode()
+        {
+            int result = Length;
+            foreach (char c in GetChars())
+                result = Hash.Combine((int)c, result);
+
+            return result;
+        }
+
+        /// <summary>
+        /// A rope that wraps a simple string.
+        /// </summary>
+        private class StringRope : Rope
+        {
+            private readonly string _value;
+            public StringRope(string value) => _value = value;
+            public override string ToString() => _value;
+            public override int Length => _value.Length;
+            protected override IEnumerable<char> GetChars() => _value;
+        }
+
+        /// <summary>
+        /// A rope that represents the concatenation of two ropes.
+        /// </summary>
+        private class ConcatRope : Rope
+        {
+            private readonly Rope _left, _right;
+            public override int Length { get; }
+
+            public ConcatRope(Rope left, Rope right)
+            {
+                _left = left;
+                _right = right;
+                Length = checked(left.Length + right.Length);
+            }
+
+            public override string ToString()
+            {
+                var psb = PooledStringBuilder.GetInstance();
+                var stack = new Stack<Rope>();
+                stack.Push(this);
+                while (stack.Count != 0)
+                {
+                    switch (stack.Pop())
+                    {
+                        case StringRope s:
+                            psb.Builder.Append(s.ToString());
+                            break;
+                        case ConcatRope c:
+                            stack.Push(c._right);
+                            stack.Push(c._left);
+                            break;
+                        case var v:
+                            throw ExceptionUtilities.UnexpectedValue(v.GetType().Name);
+                    }
+                }
+
+                return psb.ToStringAndFree();
+            }
+
+            protected override IEnumerable<char> GetChars()
```

Comment:
Hmm, this method is going to make it hard to have a `Span<char>` method here :frown: